compiler/modules/front/const.bal

```diff
@@ -517,47 +520,88 @@ function foldVarRefExpr(FoldContext cx, t:SemType? expectedType, s:VarRefExpr ex
     }
 }
 
-function foldFloatLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    // This will need to change when we support decimal
-    float|error result = floatFromDecimalLiteral(expr.untypedLiteral);
-    if result is float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
+function foldFpLiteralExpr(FoldContext cx, t:SemType? expectedType, s:FpLiteralExpr expr) returns s:ConstValueExpr|FoldError { 
+    float|decimal result;
+    t:Context tx = cx.typeContext();
+    s:FpTypeSuffix? typeSuffix = expr.typeSuffix;
+    string untypedLiteral = expr.untypedLiteral;
+    Position startPos = expr.startPos;
+    if typeSuffix != () {
+        result = typeSuffix is s:FLOAT_TYPE_SUFFIX ? floatFromFpLiteral(untypedLiteral) : check decimalFromFpLiteral(cx, untypedLiteral, startPos);
     }
     else {
-        return cx.semanticErr("invalid float literal", cause=result, pos=expr.startPos);
+        if expectedType == () || expectsFloat(tx, expectedType) {
+            result = floatFromFpLiteral(untypedLiteral);
+        }
+        else if expectsDecimal(tx, expectedType) {
+            result = check decimalFromFpLiteral(cx, untypedLiteral, startPos);
+        }
+        else {
+            result = floatFromFpLiteral(untypedLiteral);
+        }
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
 }
 
 function foldIntLiteralExpr(FoldContext cx, t:SemType? expectedType, s:IntLiteralExpr expr) returns s:ConstValueExpr|FoldError {
-    float|int|error result;
-    string ty;
-    if expr.base == 10 && expectsFloat(expectedType) {
-        result = floatFromDecimalLiteral(expr.digits);
-        ty = "float"; 
+    int|float|decimal result;
+    t:Context tx = cx.typeContext();
+    Position startPos = expr.startPos;
+    if expectedType == () || expectsInt(tx, expectedType) {
+        result = check intFromLiteral(cx, expr);
+    }  
+    else if expectsFloat(tx, expectedType) {
+        result = floatFromFpLiteral(expr.digits);
+    }
+    else if expectsDecimal(tx, expectedType) {
+        result = check decimalFromFpLiteral(cx, expr.digits, startPos);
     }
     else {
-        result = s:intFromIntLiteral(expr.base, expr.digits);
-        ty = "int";
-    }
-    if result is int|float {
-        return { startPos: expr.startPos, endPos: expr.endPos, value: result };
-    }
-    else {
-        return cx.semanticErr("invalid " + ty + " literal", cause=result, pos=expr.startPos);
+        result = check intFromLiteral(cx, expr);
     }
+    return { startPos: startPos, endPos: expr.endPos, value: result };
+}
+
+function expectsInt(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:INT));
+}
+
+function expectsFloat(t:Context cx, t:SemType semType) returns boolean {
+    return !t:isEmpty(cx, t:intersect(semType, t:FLOAT));
```

Comment:
original code was `t:isSubtypeSimple(t:intersect(semType, t:union(t:FLOAT, t:INT)), t:FLOAT)` why did you change it?

Reply:
Yes, it was implemented initially based on the 4th bullet point in https://ballerina.io/spec/lang/2021R1/#section_6.5. Now it should be changed according to https://github.com/ballerina-platform/ballerina-spec/issues/943#issuecomment-939932254.
I have tried to refactor both methods `foldFpLiteralExpr` and `foldIntLiteralExpr` according to this.